File: cloudflare_dyndns/cache.py
import ssl
from pathlib import Path
from typing import Dict, Optional, Tuple, Union

import truststore
from pydantic import BaseModel

from . import printer
from .types import IPAddress

ssl_context = truststore.SSLContext(ssl.PROTOCOL_TLS_CLIENT)


class InvalidCache(Exception):
    """Raised when we can't read the cache.
    It's either corrupted, an older version or unreadable.
    """


class ZoneRecord(BaseModel):
    zone_id: str
    record_id: str
    proxied: bool = False


class IPCache(BaseModel):
    address: Optional[IPAddress] = None
    updated_domains: Dict[str, ZoneRecord] = dict()

    def clear(self):
        self.address = None
        self.updated_domains = dict()


class Cache(BaseModel):
    ipv4: IPCache = IPCache()
    ipv6: IPCache = IPCache()

    def is_empty(self):
        return self == self.__class__()


class CacheManager:
    def __init__(
        self, cache_path: Union[str, Path], force: bool = False, *, debug: bool = False
    ):
        self._path = Path(cache_path).expanduser()
        self._force = force
        self._debug = debug

    def ensure_path(self):
        if self._debug:
            printer.info(f"Creating cache directory: {self._path}")
        if not self._path.parent.exists():
            self._path.parent.mkdir(parents=True)

    def load(self) -> Tuple[Cache, Cache]:
        new_cache = Cache()

        if self._force:
            printer.warning("Forced update, ignoring cache")
            return Cache(), new_cache

        try:
            return self._load(), new_cache
        except InvalidCache:
            self.delete()
            return Cache(), new_cache

    def _load(self):
        printer.info(f"Loading cache from: {self._path}")
        try:
            cache_json = self._path.read_text()
            cache = Cache.model_validate_json(cache_json)
        except FileNotFoundError:
            printer.info("Cache file not found.")
            return Cache()
        except Exception:
            message = "Invalid cache file"
            if self._debug:
                message += ": {cache_json}"
            printer.warning(message)
            raise InvalidCache

        if self._debug:
            printer.info(f"Loaded cache: {cache}")
        return cache

    def save(self, cache: Cache):
        cache_json = cache.model_dump_json(indent=2)
        if self._debug:
            printer.info(f"Saving cache: {cache_json}")
        printer.info(f"Saving cache to: {self._path}")
        self.ensure_path()
        self._path.write_text(cache_json)

    def delete(self):
        printer.warning(f"Deleting cache at: {self._path}")
        self._path.unlink(missing_ok=True)

File: cache.go
package main

import (
	"encoding/json"
	"fmt"
	"net"
	"os"
	"path/filepath"
)

type ZoneRecord struct {
	ZoneID   string `json:"zone_id"`
	RecordID string `json:"record_id"`
	Proxied  bool   `json:"proxied"`
}

type IPCache struct {
	Address        *net.IP               `json:"address,omitempty"`
	UpdatedDomains map[string]ZoneRecord `json:"updated_domains"`
}

func NewIPCache() *IPCache {
	return &IPCache{
		UpdatedDomains: make(map[string]ZoneRecord),
	}
}

func (c *IPCache) Clear() {
	c.Address = nil
	c.UpdatedDomains = make(map[string]ZoneRecord)
}

type Cache struct {
	IPv4 *IPCache `json:"ipv4"`
	IPv6 *IPCache `json:"ipv6"`
}

func NewCache() *Cache {
	return &Cache{
		IPv4: NewIPCache(),
		IPv6: NewIPCache(),
	}
}

func (c *Cache) IsEmpty() bool {
	return len(c.IPv4.UpdatedDomains) == 0 && len(c.IPv6.UpdatedDomains) == 0 &&
		c.IPv4.Address == nil && c.IPv6.Address == nil
}

type CacheManager struct {
	path  string
	force bool
	debug bool
}

func NewCacheManager(cachePath string, force bool, debug bool) *CacheManager {
	return &CacheManager{
		path:  cachePath,
		force: force,
		debug: debug,
	}
}

func (cm *CacheManager) ensurePath() error {
	if cm.debug {
		Info(fmt.Sprintf("Creating cache directory: %s", cm.path))
	}
	dir := filepath.Dir(cm.path)
	return os.MkdirAll(dir, 0755)
}

func (cm *CacheManager) Load() (*Cache, *Cache, error) {
	newCache := NewCache()

	if cm.force {
		Warning("Forced update, ignoring cache")
		return NewCache(), newCache, nil
	}

	oldCache, err := cm.load()
	if err != nil {
		cm.Delete()
		return NewCache(), newCache, nil
	}

	return oldCache, newCache, nil
}

func (cm *CacheManager) load() (*Cache, error) {
	Info(fmt.Sprintf("Loading cache from: %s", cm.path))

	data, err := os.ReadFile(cm.path)
	if err != nil {
		if os.IsNotExist(err) {
			Info("Cache file not found.")
			return NewCache(), nil
		}
		return nil, fmt.Errorf("failed to read cache file: %w", err)
	}

	var cache Cache
	if err := json.Unmarshal(data, &cache); err != nil {
		message := "Invalid cache file"
		if cm.debug {
			message += fmt.Sprintf(": %s", string(data))
		}
		Warning(message)
		return nil, fmt.Errorf("invalid cache format: %w", err)
	}

	if cm.debug {
		cacheJSON, _ := json.MarshalIndent(cache, "", "  ")
		Info(fmt.Sprintf("Loaded cache: %s", string(cacheJSON)))
	}

	return &cache, nil
}

func (cm *CacheManager) Save(cache *Cache) error {
	cacheJSON, err := json.MarshalIndent(cache, "", "  ")
	if err != nil {
		return fmt.Errorf("failed to marshal cache: %w", err)
	}

	if cm.debug {
		Info(fmt.Sprintf("Saving cache: %s", string(cacheJSON)))
	}

	Info(fmt.Sprintf("Saving cache to: %s", cm.path))

	if err := cm.ensurePath(); err != nil {
		return fmt.Errorf("failed to create cache directory: %w", err)
	}

	return os.WriteFile(cm.path, cacheJSON, 0644)
}

func (cm *CacheManager) Delete() {
	Warning(fmt.Sprintf("Deleting cache at: %s", cm.path))
	os.Remove(cm.path)
}
